### PR TITLE
extend password rotator to support multiple portals

### PR DESCRIPTION
**Pull request by leslie-corbalt**
https://jiraent.cms.gov/browse/CMCSMACD-291

Tested:
manually uploaded file, `MACFIN_nmalik.xlsx` to `s3://bharvey-test-same-account-bucket`
commented out function `resetPasswords` in main.go
ran script configured for 3 portal environments: dev, val, prod
for each env:
- script uploaded file to bucket after sync `automated sheet` and after processing the `MACFin sheet`
after script finished:
- downloaded the file and manually verified that the `automated sheet` in each environment was correct (synched, sorted, lowercase, timestamp set to RotateNow)

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -58,13 +70,22 @@ func portalClient() *http.Client {
...
-	automatedSheet := input.AutomatedSheetName
+func resetPasswords(f *excelize.File, input *Input, portal *Portal, s3Client S3ClientAPI, env Environment) (err error) {
+	automatedSheet := input.SheetGroups[env].AutomatedSheetName
 	rows, err := f.GetRows(automatedSheet)
 	if err != nil {
 		return err
 	}
 
+	sheetName := input.SheetGroups[env].SheetName
+	mcFinUsersToPasswordRow, err := getMACFinUsers(f, input, env)
```
Comment (hundt-corbalt) [inline main.go:81]: check this error

File: main.go [hunk 2]
```diff
@@ -58,13 +70,22 @@ func portalClient() *http.Client {
...
+	mfRows, err := f.GetRows(sheetName)
+	if err != nil {
+		return err
+	}
+	headerNameToXCoord := getHeaderToXCoord(mfRows[0])
```
Comment (hundt-corbalt) [inline main.go:86]: what does this code (and this program generally) do if one of the headers given is missing in one of the input sheets? does it fail with an error? or does it overwrite stuff in an arbitrary column? maybe we should have a test case to verify. I think the ideal behavior is that it does a check at the beginning of the program.

File: main.go (hunk 1 above)
Comment (leslie-corbalt) [reply main.go:81]: Done.

File: main.go (hunk 2 above)
Comment (leslie-corbalt) [reply main.go:86]: I added a validateMACFinSheets() function immediately after the file is downloaded. This checks that the MACFin sheets exist and that each MACFin sheet has the expected heading in row 0. I am not checking the PasswordManager sheets since those sheets are protected and cannot be tampered with.

Comment (hundt-corbalt) [reply main.go:86]: I think we should check both. e.g. what if the automated spreadsheet format changes and you try to run the program with the old parameters?

Comment (leslie-corbalt) [reply main.go:86]: I added a function that validates the automated and MACFin sheets for each portal. I added a test case to verify each check: empty sheet, wrong number of cols, missing col.

File: excel.go [hunk 3]
```diff
@@ -15,10 +15,111 @@ type PasswordRow struct {
...
+}
+
+func validateSheets(f *excelize.File, input *Input) error {
+	sheetList := f.GetSheetList()
+	for _, group := range input.SheetGroups {
+		sheets := []string{group.SheetName, group.AutomatedSheetName}
+		for _, sheet := range sheets {
+			// check that sheet exists
+			if !contains(sheetList, sheet) {
+				log.Printf("sheet %s missing from file s3://%s/%s", sheet, input.Bucket, input.Key)
```
Comment (hundt-corbalt) [inline excel.go:99]: Instead of logging details and returning a less-informative error, I think it would be better if the error message included the full details.

I don't think you need to return constant errors or special error types. If you want to check if the error matches what is expected, you can just create an error with the same text for the test case.

File: main_test.go [hunk 4]
```diff
@@ -273,12 +274,19 @@ type MACFinRow struct {
...
+	Name                                    string
+	PasswordManagerIn                       []PasswordManagerRow
+	PasswordManagerOut                      []PasswordManagerRow
+	MACFinIn                                []MACFinRow
+	UntrackedPasswords                      map[string]string // user -> password
+	ServerErrors                            map[string]string // user -> path
+	InvalidMACFinInUsernameHeading          bool
+	InvalidMACFinInPasswordHeading          bool
+	MACFinInEmpty                           bool
+	PasswordManagerInEmpty                  bool
```
Comment (hundt-corbalt) [inline main_test.go:286]: nothing happens if you set this

File: main_test.go [hunk 5]
```diff
@@ -273,12 +274,19 @@ type MACFinRow struct {
...
-	MACFinIn           []MACFinRow
-	UntrackedPasswords map[string]string // user -> password
-	ServerErrors       map[string]string // user -> path
+	Name                                    string
+	PasswordManagerIn                       []PasswordManagerRow
+	PasswordManagerOut                      []PasswordManagerRow
+	MACFinIn                                []MACFinRow
+	UntrackedPasswords                      map[string]string // user -> password
+	ServerErrors                            map[string]string // user -> path
+	InvalidMACFinInUsernameHeading          bool
```
Comment (hundt-corbalt) [inline main_test.go:283]: I like this solution to the question of how to specify a heading issue in the context of these test cases.

One thing I would suggest is, instead of having all these bools, have something like

```
type SheetProblem int
const (
  NoSheetProblem SheetProblem = iota
  SheetProblemInvalidMACFinUsernameHeading  # remove "In" from the names
  SheetProblemPasswordManagerEmpty
  // etc.
)

type TestCase struct {
  SheetInProblem SheetProblem
}

```

This makes it clearer that the test code does not support arbitrary combinations of problems, and also lets you write `tc.SheetProblem != NoSheetProblem` to see if there is a validation error expected, instead of the long disjunction of all the different booleans.

File: main_test.go [hunk 6]
```diff
@@ -734,36 +838,48 @@ func TestRotate(t *testing.T) {
...
 
 				if err != nil {
-					t.Fatalf("Error running rotate(): %s", err)
+					if errors.Is(err, ErrSheetDoesNotExist) || errors.Is(err, ErrSheetIsEmpty) || errors.Is(err, ErrSheetMissingHeader) ||
+						errors.Is(err, ErrWrongNumberOfCols) || errors.Is(err, ErrWrongColumnHeading) {
+						t.Logf("Error in input file: %s", err)
+						return
+					} else {
+						t.Fatalf("Error running rotate(): %s", err)
+					}
```
Comment (hundt-corbalt) [inline main_test.go:882]: you also need to fail the test case if there is _no_ error but there was an invalid input.

File: excel.go (hunk 3 above)
Comment (leslie-corbalt) [reply excel.go:99]: "I don't think you need to return constant errors or special error types. If you want to check if the error matches what is expected, you can just create an error with the same text for the test case." 
This is exactly what I was wondering about. You answered my unspoken question. Thank you.

File: excel.go [hunk 7]
```diff
@@ -15,10 +15,111 @@ type PasswordRow struct {
...
+			return err
+		}
+		header := rows[0]
+		// check number of cols
+		if len(header) != len(input.AutomatedSheetColNameToIndex) {
+			log.Printf("Expected sheet %s to have %d cols; it has %d cols", sheetName, len(input.AutomatedSheetColNameToIndex), len(header))
+			return ErrWrongNumberOfCols
+		}
+		// check col headings and indexes
+		if header[ColUser] != ColUserHeading {
```
[the remark this answers is not in the document]
Comment (leslie-corbalt) [reply excel.go:70]: Yes. I fixed this. I added a map to Input: `input.AutomatedSheetColNameToHeading` . I iterate through `input.AutomatedSheetColNameToIndex` and for each `col, index`, I check the heading: `header[index]` equals `input.AutomatedSheetColNameToHeading[col]`
This means if the cols or headings change, this validation code works.

File: excel.go [hunk 8]
```diff
@@ -15,10 +15,111 @@ type PasswordRow struct {
 	Row      int
 }
 
+var ErrSheetDoesNotExist = errors.New("sheet does not exist")
```
[the remark this answers is not in the document]
Comment (leslie-corbalt) [reply excel.go:18]: Yes. Thanks for your comment below. I removed all these error types.

File: main_test.go (hunk 6 above)
Comment (leslie-corbalt) [reply main_test.go:882]: yup!

File: main_test.go (hunk 4 above)
Comment (leslie-corbalt) [reply main_test.go]: I added the test for this.

File: main_test.go (hunk 5 above)
Comment (leslie-corbalt) [reply main_test.go]: Much better. Much cleaner. Much easier to understand.

File: main_test.go [hunk 9]
```diff
@@ -734,36 +838,48 @@ func TestRotate(t *testing.T) {
...
-				err = rotate(input, portal, fc)
+				err = rotate(input, envToPortal, fc)
 				server.Shutdown(context.Background())
 
 				if err != nil {
-					t.Fatalf("Error running rotate(): %s", err)
+					if errors.Is(err, ErrSheetDoesNotExist) || errors.Is(err, ErrSheetIsEmpty) || errors.Is(err, ErrSheetMissingHeader) ||
+						errors.Is(err, ErrWrongNumberOfCols) || errors.Is(err, ErrWrongColumnHeading) {
+						t.Logf("Error in input file: %s", err)
+						return
```
[the remark this answers is not in the document]
Comment (leslie-corbalt) [reply main_test.go]: If err is nil, I check whether I expected an error. If yes, then the test fails.

File: excel.go [hunk 10]
```diff
@@ -48,37 +126,34 @@ func getHeaderToXCoord(headerRow []string) map[string]int {
 	return headerToXCoord
 }
 
-type usernameAndPassword struct {
-	Username, Password string
-}
-
-func getMACFinUsers(f *excelize.File, input *Input) ([]usernameAndPassword, error) {
-	rows, err := f.GetRows(input.SheetName)
+func getMACFinUsers(f *excelize.File, input *Input, env Environment) (map[string]PasswordRow, error) {
```
[the remark this answers is not in the document]
Comment (leslie-corbalt) [reply excel.go:129]: I discussed with Nazib. He said that it is OK if the last username/password is used.

File: main.go [hunk 11]
```diff
@@ -21,20 +22,34 @@ const (
...
-	SheetName                    string
-	UsernameHeader               string
-	PasswordHeader               string
-	AutomatedSheetPassword       string
-	AutomatedSheetName           string // sheet managed by application
-	AutomatedSheetColNameToIndex map[Column]int
-	RowOffset                    int // number of header rows (common to all sheets)
+	Bucket                         string
+	Key                            string
+	SheetName                      string
```
[the remark this answers is not in the document]
Comment (leslie-corbalt) [reply main.go:45]: Yes, since SheetName is now a field in SheetGroup it is no longer necessary. Done.

File: excel.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (leslie-corbalt) [reply excel.go:203]: I added a function to remove duplicate MACFin users; in the case of duplicates, it saves the last user and deletes the others.

[the remark this answers is not in the document]
Comment (leslie-corbalt) [reply excel.go:203]: I found a lot of small things after setting rowOffset to a value > 1. By setting it to `2`:
- I discovered a few places that were referencing a hardcoded value, rather than rowOffset
- the combination of rowOffset > 1 and an empty sheet caused the program to panic (slice out of bounds)
- - I now check for this case in the `validateSheets` function
- - I also added a few test cases for rowOffset > 1

[the remark this answers is not in the document]
Comment (leslie-corbalt) [reply excel.go:203]: These changes will be put into a separate PR.

File: main_test.go [hunk 12]
```diff
@@ -995,7 +997,9 @@ func TestRotate(t *testing.T) {
 				macFinUsers := map[string]struct{}{}
 				for _, row := range tc.MACFinIn {
 					if row.Username != "" {
-						macFinUsers[row.Username] = struct{}{}
+						if _, ok := macFinUsers[strings.ToLower(row.Username)]; !ok {
```
Comment (hundt-corbalt) [inline main_test.go:1000]: why this change?

**Review by hundt-corbalt — COMMENTED** [review]
This new code doesn't test that duplicates are removed and that the names are converted to lowercase. It just checks that no names are missing and that, however many entries there are for each name, all the passwords are right. I think that's fine as long as we don't care if the duplicate-removing and the lowercasing stops happening. What do you think?

File: main_test.go (hunk 12 above)
Comment (leslie-corbalt) [reply main_test.go:1000]: This change is needed because the code now removes duplicates from the MACFin sheet, and then sets usernames to lowercase.
If I revert this change then the test on main_test.go:1031 fails because `Users are now missing from the MACFin sheet`.

Comment (hundt-corbalt) [reply main_test.go:1000]: ah, so you need the lowercase, but I still don't see why you need the `if`

**Comment by leslie-corbalt** [comment]
> This new code doesn't test that duplicates are removed and that the names are converted to lowercase. It just checks that no names are missing and that, however many entries there are for each name, all the passwords are right. I think that's fine as long as we don't care if the duplicate-removing and the lowercasing stops happening. What do you think?

If duplicate-removing stops then this test will fail, since the code assumes there is only one occurrence of each name. If there  are 2 occurrences of a name, and dups are not removed, then the code only updates the last occurrence of the name. 

Do I need another test to ensure there are no duplicates?

File: main_test.go (hunk 12 above)
Comment (leslie-corbalt) [reply main_test.go:1000]: Yup, you are correct. I don't need the `if` statement. Done.

**Review by hundt-corbalt — APPROVED** [review]
LGTM!

**Comment by leslie-corbalt** [comment]
#8
